CMakeLists.txt

```diff
@@ -587,6 +588,111 @@ if (BUILD_TESTING)
             set_property(TEST ${test_target} PROPERTY TIMEOUT 7200)
         endforeach()
     endif()
+
+    if(S2N_FUZZ_TEST)
+        set(SCRIPT_PATH "${CMAKE_CURRENT_SOURCE_DIR}/tests/fuzz/runFuzzTest.sh")
+        file(GLOB FUZZ_TEST_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/tests/fuzz/*.c")
+
+        file(GLOB TESTLIB_SRC "tests/testlib/*.c")
+        file(GLOB TESTLIB_HEADERS "tests/testlib/*.h" "tests/s2n_test.h")
+
+        add_library(fuzztest SHARED ${TESTLIB_HEADERS} ${TESTLIB_SRC})
+        target_include_directories(fuzztest PUBLIC tests)
+        target_link_libraries(fuzztest PUBLIC ${PROJECT_NAME})
+
+        # Set default values for fuzzing if not defined
+        if(NOT DEFINED LIBFUZZER_LIB)
+            set(LIBFUZZER_LIB "/usr/local/libfuzzer/lib/libFuzzer.a")
+        endif()
+        
+        if(NOT DEFINED FUZZ_TIMEOUT_SEC)
+            set(FUZZ_TIMEOUT_SEC 120)
+        endif()
+
+        if(NOT DEFINED CORPUS_UPLOAD_LOC)
+            set(CORPUS_UPLOAD_LOC "none")
+        endif()
+
+        if(NOT DEFINED ARTIFACT_UPLOAD_LOC)
+            set(ARTIFACT_UPLOAD_LOC "none")
+        endif()
+
+        if(NOT DEFINED FUZZ_TESTS)
+            set(FUZZ_TESTS "${TESTS}")
+        endif()
+
+        # Build LD_PRELOAD shared libraries
+        include_directories(
+            ${CMAKE_CURRENT_SOURCE_DIR}
+            ${CMAKE_CURRENT_SOURCE_DIR}/api
+        )
+        set(LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/tests/fuzz/LD_PRELOAD)
+        file(GLOB LIBRARY_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/tests/fuzz/LD_PRELOAD/*.c")
+        foreach(SRC ${LIBRARY_SRCS})
+            get_filename_component(LIB_NAME ${SRC} NAME_WE)
+            add_library(${LIB_NAME} SHARED ${SRC})
+
+            # Set the output directory and remove the default "lib" prefix
+            set_target_properties(${LIB_NAME} PROPERTIES
+                PREFIX ""
+                LIBRARY_OUTPUT_DIRECTORY ${LIBRARY_OUTPUT_DIRECTORY}
+            )
+            target_compile_options(${LIB_NAME} PRIVATE
+                -Wno-unreachable-code -O0 -fPIC -Wno-deprecated-declarations -fvisibility=default
```

Comment:
The following options seems a little suspicious to me. Can we either remove them or add a comment explaining why they are necessary?
- `-Wno-unreachable-code`: Can we fix the code?
- `-O0`: I notice that we do this quite a bit in our fuzz tests. Why? 🤔 
- `-Wno-deprecated-declarations`: Can we fix the code?
- `-fvisibility=default`: If we're just using the default, does it need to be explicitly specified?

And in the block below (line 653+) I'm also a bit skeptical of
- `-Wno-deprecated`
- `-Wno-unused-result`: For this one, I generally prefer explicitly ignoring the result using something like https://github.com/aws/s2n-tls/blob/9cdbea84ccdb2755fd84f405cfed66ad17a01487/utils/s2n_result.c#L98

If it ends up being more than a few lines to fix this, it might be neater to keep the fixes separate from this PR 😅